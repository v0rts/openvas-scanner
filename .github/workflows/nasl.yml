name: "NASL"

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]

jobs:
  nasl-test:
    name: Test
    runs-on: ubuntu-latest
    container: ${{ github.repository }}-build:unstable
    steps:
      - uses: actions/checkout@v3
      - name: build
        run: |
          cmake -Bbuild -DCMAKE_BUILD_TYPE=Release
          cmake --build build
      - name: redis
        run: |
          apt-get update && apt-get install --no-install-recommends --no-install-suggests -y redis
          mkdir /run/redis-openvas
          redis-server config/redis-openvas.conf || exit 1
      - name: run
        run: |
          mkdir -p /etc/openvas
          echo "db_address = /run/redis-openvas/redis.sock" >> /etc/openvas/openvas.conf
          cd nasl/tests
          OPENVAS_NASL=../../build/nasl/openvas-nasl make check
