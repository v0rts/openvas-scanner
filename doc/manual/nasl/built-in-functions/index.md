# Built-in Functions

## GENERAL

The NASL language contains a collection of built-in functions to provide many tools to detect vulnerabilities of a network device.