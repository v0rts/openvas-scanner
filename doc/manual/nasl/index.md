# NASL - NASL Attack Scripting Language

## GENERAL

The NASL Attack Scripting Language (NASL) is a language which is part of the OpenVAS Project and interpreted by the NASL interpreter. It is a simple language with focus on detecting vulnerabilities on network devices. Therefore it provides many built-in functions to attack hosts to discover vulnerabilities. A NASL script can be either run directly with the NASL interpreter [openvas-nasl](openvas-nasl.md) or within a scan with [openvas](../openvas/openvas.md).
