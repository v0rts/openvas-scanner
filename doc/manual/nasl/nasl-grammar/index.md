# NASL Grammar

## Preliminary remarks

- A comment starts with a # and finishes at the end of the current line. It is ignored by the lexical analyzer.
- “Blanks” may be inserted anywhere between two lexical tokens. 
A blank may be a sequence of white space, horizontal or vertical tabulation, line feed, form feed or carriage return characters; or a comment.
- Token are parsed by a lexical analyzer and returned to the parser.
  - As the lexical analyzer returns the longer token it finds, expressions like `a+++++b` without any white space are erroneous because they will be interpreted as `a++ ++ + b`, i.e., `(a++ ++) + b` just like in ANSI C .
  - You have to insert spaces: `a++ + ++b`
  - You cannot insert spaces in the middle of multiple character tokens, e.g., `x = a + +;` will not parse. Write `x = a ++;`
  
## Syntax

- a code block / an expression ends with a semi-colon: expr;
- function declaration: **function** identifier ( arguments ) { block }
- function call: **identifier** ( arguments )
- if block: **if** ( expr ) { block } **else if** (expr ) { block } **else** { block }
- for loop: **for** ( func; expr; func; ) { block }
- while loop: **while** ( expr ) { block }
- repeat loop: **repeat** instr **until** expr
- foreach loop: **foreach** identifier ( array ) { block }
- include inc file: **include**( "file.inc" )
- define variable: **var** = string, int, ...
- local variable: **local_var** variable
- global variable: **global_var** variable

## Types

NASL handles the following data types:

- **Integers** Any sequence of digits with an optional minus sign is an integer. NASL uses the C syntax: octal numbers can be enter by starting with 0 and hexadecimal with 0x (i.e., 0x10 = 020 = 16).

- **Strings**
  - Can be declared with single or double quotes: "string", 'string'
  - "Impure" strings
    - Entered between double quotes
    - Not converted (backslashes remain backslashes)
    - Transformed into "pure" strings by internal string function
  - "Pure" strings
    - Returned by string function or entered between simple quotes
- **Arrays**
Can be indexed with integers or strings
- **NULL** value
  Result if an initialized value is read or return of internal function in case of severe an error
  **Warnings about the NULL value**
  NULL and the array operator Reading an array element from a NULL value will immediately convert it into an array. An empty array of course, but no more an undefined variable. Changing this means big modifications in the NASL interpreter. For example:
  ```
  v = NULL;
  # isnull(v)=TRUE and typeof(v)="undef"
  x = v[2];
  # isnull(x)=TRUE and typeof(x)="undef"
  # But isnull(v)=FALSE and typeof(v)="array"
  ```

  NULL and isnull If it should be checked whether a variable is undefined, isnull(var) has to be used. Testing the equality with the NULL constant (var == NULL) is not a good idea as NULL will be converted to 0 or the empty string "" according to the type of the variable. This is necessary to ensure that variables are "automatically initialized" – changing this would probably break some existing scripts.
- **Booleans**
  Not a standalone type, comparison operators return 0 for FALSE and 1 for TRUE
  Any other value is converted:
  - The undefined or null value is FALSE
  - Integers are TRUE if not null, 0 is FALSE
  - Strings are TRUE if not empty an not "0"
  - Arrays are always TRUE whether they are empty or not


## Operators
### General Operators

- = is the assignment operator.

  - `x=42`; puts 42 into the variable x. The previous value is forgotten.

  - `x=y`; copies the value of variable y into x. If y was undefined, x becomes undefined too.

- `[ ]` is the array index operator.

  - A variable cannot be atomic and an array at the same time. If the type is changed, the previous value(s) is/are lost.
  - However, this operator can be used to extract a character from a string: if `s = "abcde"`, then `s[2] = "c"`.
  In NASL1, this could be used to change the character too: one could write `s[2] = "C";` and s became `"abCde"`. This is no longer true; the insstr function has to be used and something like `s = insstr(s, "C", 2, 2);` has to be written.
  - y[1] = 42; makes an array out of y and puts 42 in the second element. If y was not an array, it’s first undefined.

### Arithmetics Operators

Note: There is no strict rule on the integer size in NASL. The interpreter implements them with the native "int" C type, which is 32 bit long on most systems, and maybe 64 bit long on some . There is no overflow or underflow protection.

- `+` is the addition operator.
- `-` is the subtraction operator.
- `*` is the multiplication operator.
- `/` is the integer division operator.
  Note:
   - NASL does not support floating point operations.
   - Division by zero will return 0 instead of crashing the interpreter.
- `%` is the modulo. Once again, if the second operand is null, the interpreter will return 0 instead of crashing on SIGFPE.
- `**` is the exponentiation or power function.

### C Operators

NASL imported some operators from C:

- `++` is the pre-incrementation (`++x`) or post-incrementation (`x++`).
- `++x` adds 1 to x and returns the result; `x++` adds 1 to x but returns the previous value.
- `-` is the pre-decrementation (`–x`) or post-decrementation (`x–`).
- `+= -= *= /= %=` have the same meaning as in C, e.g., `x += y;` is equivalent to `x = x + y;` but x is evaluated only once. This is important in expressions like `a[i++] *= 2;` where the index “i” is incremented only once.
- `<<=` and `>>=` also exist; we added `>>>=`

### String Operators

- `+` is the string concatenation. However, you should better use the string function.
- `-` is the “string subtraction”. It removes the first instance of a string inside another.
  For example `’abcd’ - ’bc’` will give `’ad’`.
- `[]` extracts one character from the string, as explained before.
- `><` is the “string match” operator. It looks for substrings inside a string.
  `’ab’ >< ’xabcdz’` is TRUE; `’ab’ >< ’xxx’` is FALSE.
- `>!<` is the “string don’t match” operator. It looks for substrings inside a string and returns the opposite as the previous operator.
  `’ab’ >!< ’xabcdz’` is FALSE; `’ab’ >!< ’xxx’` is TRUE.
- `=~` is the “regex match” operator. It is similar to a call to the internal function ereg but is quicker because the regular expression is compiled only once when the script is parsed.
  `s =~ "[ab]*x+"` is equivalent to `ereg(string:s, pattern:"[ab]*x+", icase:1)`
- `!~` is the “regex don’t match” operator. It gives the opposite result of the previous one.

Note: =~ and !~ supports the (standard extended POSIX) regex pattern. See Supported regex expression syntax for NASL Functions for more background info on the supported regex expression support.

### Compare Operators

- `==` is TRUE if both arguments are equals, FALSE otherwise.
- `!=` is TRUE if both arguments are different, TRUE otherwise.
- `>` is the “greater than” operator.
- `>=` is the “greater than or equal” operator.
- `<` is the “lesser than” operator.
- `<=` is the “lesser than or equal” operator.

### Logical Operators

- `!` is the logical “not”. TRUE if its argument is FALSE, FALSE otherwise.
- `&&` is the logical “and”. Note that if the first argument is FALSE, the second is not evaluated.
- `||` is the logical “or”. If the first argument is TRUE, the second is not evaluated.

### Bit Fields Operators

- `~` is the arithmetic “not”, the 1-complement
- `&` is the arithmetic “and”.
- `|` is the arithmetic “or”.
- `^` is the arithmetic “xor” (exclusive or).
- `<<` is the logical bit shift to the left.
- `>>` is the arithmetic/signed shift to the right.
The sign bit, if any, is propagated.
- `>>>` is the logical/unsigned shift to the right.
  The sign bit is pushed to the right and replaced by zero.

In all shift operators, the count is on the right, i.e., `x>>2` is equivalent to `x/4` and `x<<2` is `x*4`.

### Special Behavior

- break can (but should not) be used to exit from a function or the script.OB
- In case its arguments have different types, + now tries very hard to do something smart, i.e., a string concatenation, then an integer addition. It prints a warning, though, because such automatic conversion is dangerous.
  - If one of its argument is undefined, + returns the other one.
  - If one of its argument is a “pure string”, the other argument is converted to a string if necessary, and the result is a “pure string”. An “Impure string” is converted to a pure string without escape sequence interpretation, i.e., "AB\n"+’de’ gives ’AB\\nde’, i.e., “AB”, a backslash, then “nde”.
   - If one of its argument is an “impure string”, the second argument is converted to a string if necessary and the result is an “impure string”, i.e., "ABC"+2 gives "ABC2".
   - If one of its argument is an integer, the other is converted to an integer and the result is an integer.
   - In any other case, NULL is returned.
- The “magical strings” from NASL1 have been removed. In NASL1, adding a string to an integer might give an integer if the string contained only digits.
- The minus operator follows the same type conversion rules as plus.
- Using unitialized variables is bad. However, to ensure that old scripts still work, the NULL undefined value will be converted to 0 or “” according to the context (integer or string). That is why isnull has to be used to test if a variable is undefined. See “Warnings about the NULL value” in NASL Documentation.

## Precedence
| Operators                       | Associativity |
| ------------------------------- | ------------- |
| ++ --                           | None          |
| **                              | Right         |
| ~- (unary minus)                | Left          |
| !                               | Left          |
| * / %                           | Left          |
| + -                             | Left          |
| << >> >>>                       | Left          |
| &                               | Left          |
| ^                               | Left          |
| \|                              | Left          |
| < <= > >= == != <> =~ !~ >!< >< | None          |
| &&                              | Left          |
| \|\|                            | Left          |
| = += -= *= /= %= <<= >>= >>>=   | Right         |


## Loops and Control Flow

### Operators

- `for (expr1; cond; expr2) block;` is similar to the C operator and is equivalent to `expr1; while(cond) block; expr2;`
  A classical construction to count from 1 to 10 is: `for(i=1;i<=10;i++) display(i,’{\n’);`
- `foreach var (array) block;` iterates all elements in an array.
  Note that var iterates through the values stored in the array, not the indexes. If that is desired, use: foreach var (keys(array)) block;
- `while(cond) block;` executes the block as long as the condition is TRUE.
  If the condition is FALSE, the block is never executed.
- `repeat block; until (cond);` executes the blocks as long as the condition is TRUE. The block is executed at least once.
- `break` breaks the current loop and jumps at its exit.
  If you are not inside a loop, the behavior is undefined.
- `continue` jumps to the next step of the loop.
  If you are not inside a loop, the behavior is undefined.
- `return` returns a value from the current function.

## Declarations

###Variable Declarations

NASL uses global and local variables. Local variables are created in a function and stop existing as soon as the function returns. When the interpreter looks for a variable, it first searches in the current function context, then in the calling context (if any) etc., until it reaches the top level context that contains the global variables.

Normally, a variable does not have to be declared: either it exists, because it was already used in this context, or because a calling function used it, or it will be created in the current context.
However, this may be dangerous in some cases:

1. If you want to write into a global variable from within a function and cannot be sure that the variable was created first in the top level context, or created as a local variable in a calling function context.
2. If you want to be sure that you are creating a brand new local variable and not overwriting a global variable with the same name.

A variable can explicitly be declared as follows:

- local_var var;
- global_var var;

If the variable already exists in the specified context, an error message is returned.

### Function Declarations

 - function name (argname1, argname2) block;

Note: The argument list may be empty, but if it is not, user-defined function parameters must be named. Unnamed arguments may be used without being declared.

### Retrieving Function Arguments

Inside an NASL function, named arguments are just accessed as any local variable. Unnamed arguments are implemented through the special array _FCT_ANON_ARGS.
This variable will be NULL in interpreters below NASL_LEVEL 2190.

The following may be put at the start of scripts that need this function:

```
if (NASL_LEVEL < 2190) exit(0); # _FCT_ANON_ARGS is not implemented
```

1. Writing to _FCT_ANON_ARGS is undefined. Currently, the memory is wasted but the value cannot be read back.
2. Using _FCT_ANON_ARGS to try to read named arguments is bad too. Currently, there is a protection and a NULL value is returned.

### Calling Functions

Here is an example with named arguments:
```
function fact(n)
{
   local_var i, f;
   f = 1;
   for (i = 1; i <= n; i ++) f *= i;
   return f;
}
display("3 ! = ", fact(n: 3), "\n");
```

And the same with unnamed arguments:
```
function fact()
{
   local_var i, f;
   f = 1;
   for (i = 1; i <= _FCT_ANON_ARGS[0]; i ++) f *= i;
   return f;
}
display("3 ! = ", fact(3), "\n");
```

And another, mixing the two flavors:
```
function fact(prompt)
{
   local_var i, f;
   f = 1;
   for (i = 1; i <= _FCT_ANON_ARGS[0]; i ++)
   {
      f *= i;
      display(prompt, i, ’! = ’, f, ’\n’);
   }
   return f;
}
n = fact(3, prompt: ’> ’);
```


