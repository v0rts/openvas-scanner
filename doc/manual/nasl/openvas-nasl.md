# openvas-nasl

## NAME

openvas-nasl - NASL Attack Scripting Language

## SYNOPSIS

**openvas-nasl** *\<\[-Vh\] \[-T tracefile\] \[-s\] \[-t target\] \[-c
config_file\] \[-d\] \[-sX\] \> files\...*

## DESCRIPTION

**openvas-nasl** executes a set of NASL scripts against a given target
host. It can also be used to determine if a NASL script has any syntax
errors by running it in parse (**-p**) or lint (**-L**) mode.

## OPTIONS

**-T tracefile**

:   Makes nasl write verbosely what the script does in the file
    *tracefile* , ala \'set -x\' under sh

**-t target**

:   Apply the NASL script to *target* which may be a single host
    (127.0.0.1), a whole subnet (192.168.1.0/24) or several subnets
    (192.168.1.0/24, 192.168.243.0/24)

**-e iface**

:   Specifies the network interface to be used as the source for
    established connections.

**-s**

:   Sets the return value of safe_checks() to 1. (See the OpenVAS
    Scanner documentation to know what the safe checks are) Implies -B.

**-D**

:   Only run the description part of the script.

**-B**

:   Runs in description mode before running the script.

**-L**

:   **Lint** the script (run extended checks).

**-X**

:   Run the script with disabled signature verification.

**-h**

:   Show help

**-V**

:   Show the version of NASL.

**-d**

:   Output debug information to stderr.

**-r port-range**

:   This is the default range of ports that the scanner plugins will
    probe. The syntax of this option is flexible, it can be a single
    range (\"1-1500\"), several ports (\"21,23,80\"), several ranges of
    ports (\"1-1500,32000-33000\"). Note that you can specify UDP and
    TCP ports by prefixing each range by T or U. For instance, the
    following range will make openvas scan UDP ports 1 to 1024 and TCP
    ports 1 to 65535 : \"T:1-65535,U:1-1024\".

**-k key=value**

:   Set KB key to value. Can be used multiple times.

## SEE ALSO

**[openvas(1)](../openvas/openvas.md)**, **[openvas-nasl-lint(1)](openvas-nasl-lint.md)**

## HISTORY

NASL comes from a private project called \'pkt_forge\', which was
written in late 1998 by Renaud Deraison and which was an interactive
shell to forge and send raw IP packets (this pre-dates Perl\'s
Net::RawIP by a couple of weeks). It was then extended to do a wide
range of network-related operations and integrated into the scanner as
\'NASL\'.

The parser was completely hand-written and a pain to work with. In
Mid-2002, Michel Arboi wrote a bison parser for NASL, and he and Renaud
Deraison re-wrote NASL from scratch. Although the \"new\" NASL was
nearly working as early as August 2002, Michel\'s laziness made us wait
for early 2003 to have it working completely.

After the original authors decided to stop the Open Source development
in 2005, most changes and maintenance works were done by Greenbone
Networks.

## AUTHOR

Most of the engine is (C) 2003 Michel Arboi, most of the built-in
functions are (C) 2003 Renaud Deraison. Most new code since 2005
developed by Greenbone Networks GmbH.
