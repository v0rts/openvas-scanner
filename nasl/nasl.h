/* SPDX-FileCopyrightText: 2023 Greenbone AG
 * SPDX-FileCopyrightText: 2002-2005 Tenable Network Security
 *
 * SPDX-License-Identifier: GPL-2.0-only
 */

#ifndef NASL_H
#define NASL_H

#include "../misc/scanneraux.h"

#include <glib.h>

/* Signature information extraction and verification (not nasl- specific
  anymore, thus likely to be moved to openvas-libraries): */
int
nasl_verify_signature (const char *filename);

char *
nasl_extract_signature_fprs (const char *filename);

GSList *
nasl_get_all_certificates (void);
/* End of Signature information extraction */

int
add_nasl_inc_dir (const char *);

void
nasl_clean_inc (void);

int
exec_nasl_script (struct script_infos *, int);

char *
nasl_version (void);

pid_t
nasl_server_start (char *, char *);

void
nasl_server_recompile (char *, char *);

/* exec_nasl_script modes */
#define NASL_EXEC_DESCR (1 << 0)
#define NASL_EXEC_PARSE_ONLY (1 << 1)
#define NASL_ALWAYS_SIGNED (1 << 2)
#define NASL_COMMAND_LINE (1 << 3)
#define NASL_LINT (1 << 4)

#define NASL_ERR_NOERR 0
#define NASL_ERR_ETIMEDOUT 1
#define NASL_ERR_ECONNRESET 2
#define NASL_ERR_EUNREACH 3
#define NASL_ERR_EUNKNOWN 99
#endif
