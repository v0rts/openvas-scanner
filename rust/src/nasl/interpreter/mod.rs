// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]

mod error;

mod assign;
mod call;
mod code_interpreter;
mod declare;
mod include;
#[allow(clippy::module_inception)]
mod interpreter;
mod loop_extension;
mod operator;

#[cfg(test)]
mod tests;

pub use code_interpreter::*;
pub use error::FunctionError;
pub use error::InterpretError;
pub use error::InterpretErrorKind;
pub use interpreter::Interpreter;
