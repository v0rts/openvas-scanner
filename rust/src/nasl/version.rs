#[derive(Clone, Copy, Default, Debug)]
pub enum NaslVersion {
    #[default]
    V1,
    #[allow(unused)]
    V2,
}
