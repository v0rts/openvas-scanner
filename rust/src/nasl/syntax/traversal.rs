use super::{
    grammar::*,
    visitor::{Visitor, walk_ast},
};

struct ExprCollector<'a> {
    exprs: Vec<&'a Expr>,
}

impl<'a> ExprCollector<'a> {
    fn new() -> Self {
        Self { exprs: Vec::new() }
    }
}

impl<'a> Visitor<'a> for ExprCollector<'a> {
    fn visit_expr(&mut self, expr: &'a Expr) {
        self.exprs.push(expr);
    }
}

struct StmtCollector<'a> {
    stmts: Vec<&'a Statement>,
}

impl<'a> StmtCollector<'a> {
    fn new() -> Self {
        Self { stmts: Vec::new() }
    }
}

impl<'a> Visitor<'a> for StmtCollector<'a> {
    fn visit_statement(&mut self, stmt: &'a Statement) {
        self.stmts.push(stmt);
    }
}

pub struct ExprIterator<'a> {
    exprs: Vec<&'a Expr>,
    index: usize,
}

impl<'a> ExprIterator<'a> {
    pub fn new(ast: &'a Ast) -> Self {
        let mut collector = ExprCollector::new();
        walk_ast(&mut collector, ast);
        Self {
            exprs: collector.exprs,
            index: 0,
        }
    }
}

pub struct StmtIterator<'a> {
    stmts: Vec<&'a Statement>,
    index: usize,
}

impl<'a> StmtIterator<'a> {
    pub fn new(ast: &'a Ast) -> Self {
        let mut collector = StmtCollector::new();
        walk_ast(&mut collector, ast);
        Self {
            stmts: collector.stmts,
            index: 0,
        }
    }
}

impl<'a> Iterator for ExprIterator<'a> {
    type Item = &'a Expr;

    fn next(&mut self) -> Option<Self::Item> {
        if self.index < self.exprs.len() {
            let expr = self.exprs[self.index];
            self.index += 1;
            Some(expr)
        } else {
            None
        }
    }
}

impl<'a> Iterator for StmtIterator<'a> {
    type Item = &'a Statement;

    fn next(&mut self) -> Option<Self::Item> {
        if self.index < self.stmts.len() {
            let stmt = self.stmts[self.index];
            self.index += 1;
            Some(stmt)
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::nasl::Code;

    #[test]
    fn iter_exprs() {
        let ast = Code::from_string(
            r#"
function foo() {
    a += 1;
    b += 1;
}

if (a == 1) {
    a += 1;
}
else {
    b += 1;
}
            "#,
        )
        .parse()
        .emit_errors()
        .unwrap();

        let collected: Vec<_> = ast.iter_exprs().collect();
        insta::assert_debug_snapshot!(collected);
    }

    #[test]
    fn iter_stmts() {
        let ast = Code::from_string(
            r#"
function foo() {
    a += 1;
    b += 1;
}

if (a == 1) {
    a += 1;
}
else {
    b += 1;
}
            "#,
        )
        .parse()
        .emit_errors()
        .unwrap();

        let collected: Vec<_> = ast.iter_stmts().collect();
        insta::assert_debug_snapshot!(collected);
    }
}
