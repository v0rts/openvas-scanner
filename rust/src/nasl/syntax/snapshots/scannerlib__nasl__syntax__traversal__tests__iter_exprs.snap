---
source: src/nasl/syntax/traversal.rs
expression: collected
---
[
    Assignment(
        Assignment {
            lhs: PlaceExpr {
                ident: Ident {
                    ident: "a",
                    span: Span {
                        start: CharIndex(
                            22,
                        ),
                        end: CharIndex(
                            23,
                        ),
                    },
                },
                array_accesses: [],
                negate: false,
            },
            op: AssignmentOperator {
                span: Span {
                    start: CharIndex(
                        24,
                    ),
                    end: CharIndex(
                        26,
                    ),
                },
                kind: PlusEqual,
            },
            rhs: Atom(
                Literal(
                    Literal {
                        kind: Number(
                            1,
                        ),
                        span: Span {
                            start: CharIndex(
                                27,
                            ),
                            end: CharIndex(
                                28,
                            ),
                        },
                    },
                ),
            ),
        },
    ),
    Atom(
        Literal(
            Literal {
                kind: Number(
                    1,
                ),
                span: Span {
                    start: CharIndex(
                        27,
                    ),
                    end: CharIndex(
                        28,
                    ),
                },
            },
        ),
    ),
    Assignment(
        Assignment {
            lhs: PlaceExpr {
                ident: Ident {
                    ident: "b",
                    span: Span {
                        start: CharIndex(
                            34,
                        ),
                        end: CharIndex(
                            35,
                        ),
                    },
                },
                array_accesses: [],
                negate: false,
            },
            op: AssignmentOperator {
                span: Span {
                    start: CharIndex(
                        36,
                    ),
                    end: CharIndex(
                        38,
                    ),
                },
                kind: PlusEqual,
            },
            rhs: Atom(
                Literal(
                    Literal {
                        kind: Number(
                            1,
                        ),
                        span: Span {
                            start: CharIndex(
                                39,
                            ),
                            end: CharIndex(
                                40,
                            ),
                        },
                    },
                ),
            ),
        },
    ),
    Atom(
        Literal(
            Literal {
                kind: Number(
                    1,
                ),
                span: Span {
                    start: CharIndex(
                        39,
                    ),
                    end: CharIndex(
                        40,
                    ),
                },
            },
        ),
    ),
    Binary(
        Binary {
            lhs: Atom(
                Ident(
                    Ident {
                        ident: "a",
                        span: Span {
                            start: CharIndex(
                                49,
                            ),
                            end: CharIndex(
                                50,
                            ),
                        },
                    },
                ),
            ),
            op: BinaryOperator {
                span: Span {
                    start: CharIndex(
                        51,
                    ),
                    end: CharIndex(
                        53,
                    ),
                },
                kind: EqualEqual,
            },
            rhs: Atom(
                Literal(
                    Literal {
                        kind: Number(
                            1,
                        ),
                        span: Span {
                            start: CharIndex(
                                54,
                            ),
                            end: CharIndex(
                                55,
                            ),
                        },
                    },
                ),
            ),
        },
    ),
    Atom(
        Ident(
            Ident {
                ident: "a",
                span: Span {
                    start: CharIndex(
                        49,
                    ),
                    end: CharIndex(
                        50,
                    ),
                },
            },
        ),
    ),
    Atom(
        Literal(
            Literal {
                kind: Number(
                    1,
                ),
                span: Span {
                    start: CharIndex(
                        54,
                    ),
                    end: CharIndex(
                        55,
                    ),
                },
            },
        ),
    ),
    Assignment(
        Assignment {
            lhs: PlaceExpr {
                ident: Ident {
                    ident: "a",
                    span: Span {
                        start: CharIndex(
                            63,
                        ),
                        end: CharIndex(
                            64,
                        ),
                    },
                },
                array_accesses: [],
                negate: false,
            },
            op: AssignmentOperator {
                span: Span {
                    start: CharIndex(
                        65,
                    ),
                    end: CharIndex(
                        67,
                    ),
                },
                kind: PlusEqual,
            },
            rhs: Atom(
                Literal(
                    Literal {
                        kind: Number(
                            1,
                        ),
                        span: Span {
                            start: CharIndex(
                                68,
                            ),
                            end: CharIndex(
                                69,
                            ),
                        },
                    },
                ),
            ),
        },
    ),
    Atom(
        Literal(
            Literal {
                kind: Number(
                    1,
                ),
                span: Span {
                    start: CharIndex(
                        68,
                    ),
                    end: CharIndex(
                        69,
                    ),
                },
            },
        ),
    ),
    Assignment(
        Assignment {
            lhs: PlaceExpr {
                ident: Ident {
                    ident: "b",
                    span: Span {
                        start: CharIndex(
                            84,
                        ),
                        end: CharIndex(
                            85,
                        ),
                    },
                },
                array_accesses: [],
                negate: false,
            },
            op: AssignmentOperator {
                span: Span {
                    start: CharIndex(
                        86,
                    ),
                    end: CharIndex(
                        88,
                    ),
                },
                kind: PlusEqual,
            },
            rhs: Atom(
                Literal(
                    Literal {
                        kind: Number(
                            1,
                        ),
                        span: Span {
                            start: CharIndex(
                                89,
                            ),
                            end: CharIndex(
                                90,
                            ),
                        },
                    },
                ),
            ),
        },
    ),
    Atom(
        Literal(
            Literal {
                kind: Number(
                    1,
                ),
                span: Span {
                    start: CharIndex(
                        89,
                    ),
                    end: CharIndex(
                        90,
                    ),
                },
            },
        ),
    ),
]
